### PR TITLE
AEA-689 Generator Tests

### PR DESCRIPTION
**Pull request by 5A11**
This PR increases test coverage for the generator package.

This PR is a work in progress. 

It covers generator/common.py and generator/extract_specification.py modules, to the extent I knew how to write test for them. The methods for which I did not know how to tests contain a "# ToDo" mark.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/data/generator/t_protocol/__init__.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 """This module contains the support resources for the t_protocol protocol."""
 
-from tests.data.generator.t_protocol.message import TProtocolMessage
-from tests.data.generator.t_protocol.serialization import TProtocolSerializer
+from .message import TProtocolMessage
```
Comment (DavidMinarsch) [inline tests/data/generator/t_protocol/__init__.py:22]: paths should be absolute as a rule

File: tests/data/generator/t_protocol/custom_types.py [hunk 2]
```diff
@@ -67,8 +67,8 @@ def encode(data_model_protobuf_object, data_model_object: "DataModel") -> None:
         data_model_protobuf_object.list_field.extend(data_model_object.list_field)
         data_model_protobuf_object.dict_field.update(data_model_object.dict_field)
 
-    @staticmethod
-    def decode(data_model_protobuf_object) -> "DataModel":
+    @classmethod
```
Comment (DavidMinarsch) [inline tests/data/generator/t_protocol/custom_types.py:70]: this is a staticmethod in all other protocols; why the change?

Comment (DavidMinarsch) [reply tests/data/generator/t_protocol/custom_types.py:70]: You can add it as an issue to change later but not now

File: aea/protocols/generator/validate.py [hunk 3]
```diff
@@ -147,7 +147,16 @@ def _is_valid_union(content_type: str) -> bool:
...
+    # check there are at least two subtypes in the union
     sub_types = _get_sub_types_of_compositional_types(content_type)
+    if len(sub_types) < 2:
+        return False
+
+    # check there are no duplicate subtypes in the union
+    sub_types_set = set(sub_types)
+    if len(sub_types) != len(sub_types_set):
+        return False
+
```
Comment (5A11) [inline aea/protocols/generator/validate.py:159]: Small addition to the validator section as a result of writing the tests.

File: tests/data/generator/t_protocol/custom_types.py (hunk 2 above)
Comment (DavidMinarsch) [reply tests/data/generator/t_protocol/custom_types.py:70]: We also decided it to be a staticmethod a while back; so probably no issue either.

File: tests/data/sample_specification.yaml [hunk 4]
```diff
@@ -1,3 +1,4 @@
+---
```
Comment (5A11) [inline tests/data/sample_specification.yaml:1]: This test specification has been beefed up to accommodate all new test cases, including dialogue part which was previously missing.

File: tests/data/generator/t_protocol/protocol.yaml [hunk 5]
```diff
@@ -5,12 +5,13 @@ description: A protocol for testing purposes.
 license: Apache-2.0
 aea_version: '>=0.5.0, <0.6.0'
 fingerprint:
```
Comment (DavidMinarsch) [inline tests/data/generator/t_protocol/protocol.yaml:7]: where is the specification for this protocol?

Comment (5A11) [reply tests/data/generator/t_protocol/protocol.yaml:7]: `tests/data/sample_specification,yaml`

File: tests/test_protocols/test_generator.py [hunk 6]
```diff
@@ -17,16 +17,15 @@
 #
 # ------------------------------------------------------------------------------
 """This module contains the tests for the protocol generator."""
-
-import inspect
+import filecmp
 import logging
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator.py:21]: Please break up into a directory with multiple test files; following structure in aea package

File: tests/data/generator/t_protocol/__init__.py (hunk 1 above)
Comment (5A11) [reply tests/data/generator/t_protocol/__init__.py:22]: Done

File: tests/test_protocols/test_generator.py (hunk 6 above)
Comment (5A11) [reply tests/test_protocols/test_generator.py:21]: Done

File: tests/test_protocols/test_generator/test_common.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+        assert _match_brackets(text_1, 3) == 11
+        assert _match_brackets(text_1, 7) == 10
+
+        text_2 = "[]]som[]et[hi[ng][sf]"
+        index_2 = 4
+        with self.assertRaises(SyntaxError) as cm:
+            _match_brackets(text_2, index_2)
+        self.assertEqual(
+            str(cm.exception),
+            "'index_of_open_bracket' in 'text' is not an open bracket '['. It is {}".format(
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_common.py:90]: Shouldn't the error be `"Index {} in 'text' is not an open bracket '['. It is {}".format(index, what_it_is)`?

File: tests/test_protocols/test_generator/test_common.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+        content_type_not_includes_3 = (
+            "Optional[Union[int, float, FrozenSet[int], Tuple[bool, ...], float]]"
+        )
+        assert _includes_custom_type(content_type_not_includes_3) is False
+
+    def test_is_installed(self,):
+        """Test the 'is_installed' method"""
+        # ToDo
+        pass
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_common.py:319]: It;s enough to call the method and then mock the response.

File: tests/test_protocols/test_generator/test_common.py [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+    def test_check_prerequisites(self,):
+        """Test the 'check_prerequisites' method"""
+        # ToDo
+        pass
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_common.py:324]: see directly above

File: tests/test_protocols/test_generator/test_end_to_end.py [hunk 10]
```diff
@@ -83,150 +69,6 @@ def setup_class(cls):
         create_private_key(DEFAULT_LEDGER, cls.private_key_path_1)
         create_private_key(DEFAULT_LEDGER, cls.private_key_path_2)
 
-    def test_compare_latest_generator_output_with_test_protocol(self):
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_end_to_end.py:86]: Why has this been removed?

File: tests/test_protocols/test_generator/test_extract_specification.py [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+        }
+
+    @mock.patch(
+        "aea.protocols.generator.validate.validate",
+        return_value=[False, "some error."],
+    )
+    def test_extract_negative_invalid_specification(self, validate_mock):
+        """Negative test the 'extract' method."""
+        pytest.skip("todo")
+        # ToDo
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_extract_specification.py:517]: TODO should always have a comment what's todo

File: tests/test_protocols/test_generator/test_extract_specification.py [hunk 12]
```diff
@@ -0,0 +1,533 @@
...
+        expected_3 = "FrozenSet[int]"
+        assert _specification_type_to_python_type(specification_type_3) == expected_3
+
+        specification_type_4 = "pt:list[pt:float]"
+        expected_4 = "Tuple[float, ...]"
+        assert _specification_type_to_python_type(specification_type_4) == expected_4
+
+        specification_type_5 = "pt:dict[pt:bool, pt:str]"
+        expected_5 = "Dict[bool, str]"
+        assert _specification_type_to_python_type(specification_type_5) == expected_5
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_extract_specification.py:312]: This would be a great usecase for fixtures

File: tests/test_protocols/test_generator/test_end_to_end.py (hunk 10 above)
Comment (DavidMinarsch) [reply tests/test_protocols/test_generator/test_end_to_end.py:86]: I think it has been moved only

File: tests/test_protocols/test_generator/test_generator.py [hunk 13]
```diff
@@ -0,0 +1,1002 @@
...
+        # compare __init__.py
+        init_file_generated = Path(self.t, T_PROTOCOL_NAME, "__init__.py")
+        init_file_original = Path(PATH_TO_T_PROTOCOL, "__init__.py",)
+        assert filecmp.cmp(init_file_generated, init_file_original)
+
+        # protocol.yaml are different because the custom_types.py files are different
+        # # compare protocol.yaml
+        # protocol_yaml_file_generated = Path(self.t, T_PROTOCOL_NAME, "protocol.yaml")
+        # protocol_yaml_file_original = Path(PATH_TO_T_PROTOCOL, "protocol.yaml",)
+        # assert filecmp.cmp(protocol_yaml_file_generated, protocol_yaml_file_original)
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_generator.py:86]: If code is not needed remove; else describe the issue in a TODO

File: tests/test_protocols/test_generator/test_validate.py [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+)
+
+logger = logging.getLogger("aea")
+logging.basicConfig(level=logging.INFO)
+
+
+class TestValidate(TestCase):
+    """Test for generator/validate.py."""
+
+    def test_is_reserved_name(self):
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator/test_validate.py:42]: Another great usecase for fixtures!

File: tests/test_protocols/test_generator/test_common.py (hunk 7 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_common.py:90]: Done

File: tests/test_protocols/test_generator/test_end_to_end.py (hunk 10 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_end_to_end.py:86]: Yes it is moved to `test_generator.py`.

File: tests/test_protocols/test_generator/test_extract_specification.py (hunk 11 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_extract_specification.py:517]: `# ToDo: Complete this test!`

File: tests/test_protocols/test_generator/test_extract_specification.py (hunk 12 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_extract_specification.py:312]: I don't know about fixtures. I will follow up.

File: tests/test_protocols/test_generator/test_generator.py (hunk 13 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_generator.py:86]: Done

File: tests/test_protocols/test_generator/test_common.py (hunk 8 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_common.py]: Done

File: tests/test_protocols/test_generator/test_common.py (hunk 9 above)
Comment (5A11) [reply tests/test_protocols/test_generator/test_common.py]: Done
